api/staged_products_property_collection_helpers.go

```diff
@@ -0,0 +1,142 @@
+
+	if updatedPropertyAsMap, ok := updatedProperty.(map[string]interface{}); ok {
+		rawItems := updatedPropertyAsMap["value"]
+		if rawItemSlice, ok := rawItems.([]interface{}); ok {
+			for _, item := range rawItemSlice {
+				if itemMap, ok := item.(map[string]interface{}); ok {
+					collection = append(collection, updatedPropertyCollectionItem{Data: itemMap})
+				} else {
+					return nil, fmt.Errorf("parseUpdatedPropertyCollection: failed to convert %v to map[string]interface{}", item)
+				}
+			}
+		} else {
+			return nil, fmt.Errorf("parseUpdatedPropertyCollection: failed to convert %v to []interface{}", rawItems)
+		}
+	} else {
+		return nil, fmt.Errorf("parseUpdatedPropertyCollection: failed to convert %v to map[string]interface{}", updatedProperty)
+	}
+	return collection, nil
+}
+
+type responsePropertyCollection []responsePropertyCollectionItem
+type responsePropertyCollectionItem struct {
+	Data map[interface{}]interface{}
+}
+
+func parseResponsePropertyCollection(rawItems interface{}) (responsePropertyCollection, error) {
+	var collection responsePropertyCollection
+	if rawItemSlice, ok := rawItems.([]interface{}); ok {
+		for _, item := range rawItemSlice {
+			if itemMap, ok := item.(map[interface{}]interface{}); ok {
+				collection = append(collection, responsePropertyCollectionItem{Data: itemMap})
+			} else {
+				return nil, fmt.Errorf("parseResponsePropertyCollection: failed to convert %v to map[interface{}]interface{}", item)
+			}
+		}
+	} else {
+		return nil, fmt.Errorf("parseResponsePropertyCollection: failed to convert %v to []interface{}", rawItems)
+	}
+	return collection, nil
+}
+
+func (item responsePropertyCollectionItem) getFieldValue(fieldName string) string {
+	if valueObj, ok := item.Data[fieldName].(map[interface{}]interface{}); ok {
+		if value, ok := valueObj["value"].(string); ok {
+			return value
+		}
+	}
+	return ""
+}
+
+//Find and return first string matching regex pattern
+//returns matched <string, true> if found, or <"",false> if not
+func findFirst(strings []string, fieldRegex string) (string, bool) {
+	r, err := regexp.Compile(fieldRegex)
+	if err != nil {
+		panic(fmt.Sprintf("failed to compile regex: %v because:\n%v", fieldRegex, err))
+	}
+	for _, str := range strings {
+		if r.MatchString(str) {
+			return str, true
+		}
+	}
+	return "", false
+}
+
+//Finds logical key field (if exists)
+//returns <fieldName, true> if map contains field that can be considered a logical key
+//returns <"", false> if no logical key found
+func (item responsePropertyCollectionItem) findLogicalKeyField() (string, bool) {
+	//Extract & sort the fieldnames to ensure search is deterministic (since map order isn't guarenteed)
+	sortedFields := make([]string, 0, len(item.Data))
+	for k := range item.Data {
+		sortedFields = append(sortedFields, fmt.Sprintf("%v", k))
+	}
+	sort.Strings(sortedFields)
+
+	if fieldName, ok := findFirst(sortedFields, "^name$"); ok { //First look for a field named 'name'
```

Comment:
instead of a list of if/else statements that call a function used only in this method, we can create a slice (essentially an array) of the regexes, and then loop over those and call that function in the for loop. for example:
```
regexes := []string{"^name$", "^key$", "(?i)name$"}
	for _, regex := range regexes {
		compiledRegex := regexp.MustCompile(regex)

		for _, str := range sortedFields {
			if compiledRegex.MatchString(str) {
				return str
			}
		}
	}
```

Reply:
That is much better - thanks!  Implemented in 34f7985

The order in which we sort is important - [test](https://github.com/pivotal-cf/om/pull/502/commits/9e73e245a6467a586e74af526c8e3a94377db414#diff-cb697bc504ae3e69003a9c246d254be0R250) - so I left in [a comment calling this out](https://github.com/pivotal-cf/om/pull/502/commits/34f7985866373f2ddbe5b34ca3e32bb16ffc0e1d#diff-05a0e719066db6ea2a4255f414b423ebR100).